File: MyApplication/app/src/main/java/com/martinmarinkovic/myapplication/lockscreen/PinLockAdapter.java
package com.martinmarinkovic.myapplication.lockscreen;

import android.content.Context;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.recyclerview.widget.RecyclerView;

import com.martinmarinkovic.myapplication.R;

public class PinLockAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int VIEW_TYPE_NUMBER = 0;
    private static final int VIEW_TYPE_DELETE = 1;

    private Context mContext;
    private CustomizationOptionsBundle mCustomizationOptionsBundle;
    private OnNumberClickListener mOnNumberClickListener;
    private OnDeleteClickListener mOnDeleteClickListener;
    private int mPinLength;

    private int[] mKeyValues;
    private int[] mBackgrounds, mSelBackgrounds;

    public PinLockAdapter(Context context) {
        this.mContext = context;
        this.mKeyValues = getAdjustKeyValues(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 0});
        this.mBackgrounds = new int[]{R.drawable.btn_1, R.drawable.btn_2, R.drawable.btn_3, R.drawable.btn_4, R.drawable.btn_5,
                R.drawable.btn_6, R.drawable.btn_7, R.drawable.btn_8, R.drawable.btn_9, R.drawable.btn_0, R.drawable.btn_0};
        mSelBackgrounds = new int[]{R.drawable.btn_1_sel, R.drawable.btn_2_sel, R.drawable.btn_3_sel, R.drawable.btn_4_sel, R.drawable.btn_5_sel,
                R.drawable.btn_6_sel, R.drawable.btn_7_sel, R.drawable.btn_8_sel, R.drawable.btn_9_sel, R.drawable.btn_0_sel, R.drawable.btn_0_sel};
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        if (viewType == VIEW_TYPE_NUMBER) {
            View view = inflater.inflate(R.layout.layout_number_item, parent, false);
            viewHolder = new NumberViewHolder(view);
        } else {
            View view = inflater.inflate(R.layout.layout_delete_item, parent, false);
            viewHolder = new DeleteViewHolder(view);
        }
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder.getItemViewType() == VIEW_TYPE_NUMBER) {
            NumberViewHolder vh1 = (NumberViewHolder) holder;
            configureNumberButtonHolder(vh1, position);
        } else if (holder.getItemViewType() == VIEW_TYPE_DELETE) {
            DeleteViewHolder vh2 = (DeleteViewHolder) holder;
            configureDeleteButtonHolder(vh2);
        }
    }

    private void configureNumberButtonHolder(NumberViewHolder holder, int position) {
        if (holder != null) {
            if (position == 9) {
                holder.mNumberButton.setVisibility(View.GONE);
                holder.mNumberButton.setTag(mKeyValues[position]);
            } else {
                holder.mNumberButton.setBackgroundResource(mBackgrounds[position]);
                //holder.mNumberButton.setText(String.valueOf(mKeyValues[position]));
                holder.mNumberButton.setVisibility(View.VISIBLE);
                holder.mNumberButton.setTag(mKeyValues[position]);
            }

            if (mCustomizationOptionsBundle != null) {
                holder.mNumberButton.setTextColor(mCustomizationOptionsBundle.getTextColor());
                if (mCustomizationOptionsBundle.getButtonBackgroundDrawable() != null) {
                    if (Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.JELLY_BEAN) {
                        holder.mNumberButton.setBackgroundDrawable(
                                mCustomizationOptionsBundle.getButtonBackgroundDrawable());
                    } else {
                        holder.mNumberButton.setBackground(
                                mCustomizationOptionsBundle.getButtonBackgroundDrawable());
                    }
                }
                holder.mNumberButton.setTextSize(TypedValue.COMPLEX_UNIT_PX,
                        mCustomizationOptionsBundle.getTextSize());
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                        mCustomizationOptionsBundle.getButtonSize(),
                        mCustomizationOptionsBundle.getButtonSize());
                holder.mNumberButton.setLayoutParams(params);
            }
        }
    }

    private void configureDeleteButtonHolder(DeleteViewHolder holder) {
        if (holder != null) {
            if (mCustomizationOptionsBundle.isShowDeleteButton() && mPinLength > 0) {
                holder.mDeleteButton.setVisibility(View.VISIBLE);
                /*if (mCustomizationOptionsBundle.getDeleteButtonDrawable() != null) {
                    holder.mButtonImage.setImageDrawable(mCustomizationOptionsBundle.getDeleteButtonDrawable());
                }
                holder.mButtonImage.setColorFilter(mCustomizationOptionsBundle.getTextColor(),
                        PorterDuff.Mode.SRC_ATOP);
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                        mCustomizationOptionsBundle.getDeleteButtonSize(),
                        mCustomizationOptionsBundle.getDeleteButtonSize());
                holder.mButtonImage.setLayoutParams(params);*/
            } else {
                holder.mDeleteButton.setVisibility(View.GONE);
            }
        }
    }

    @Override
    public int getItemCount() {
        return 12;
    }

    @Override
    public int getItemViewType(int position) {
        if (position == getItemCount() - 1) {
            return VIEW_TYPE_DELETE;
        }
        return VIEW_TYPE_NUMBER;
    }

    public int getPinLength() {
        return mPinLength;
    }

    public void setPinLength(int pinLength) {
        this.mPinLength = pinLength;
    }

    public int[] getKeyValues() {
        return mKeyValues;
    }

    public void setKeyValues(int[] keyValues) {
        this.mKeyValues = getAdjustKeyValues(keyValues);
        notifyDataSetChanged();
    }

    private int[] getAdjustKeyValues(int[] keyValues) {
        int[] adjustedKeyValues = new int[keyValues.length + 1];
        for (int i = 0; i < keyValues.length; i++) {
            if (i < 9) {
                adjustedKeyValues[i] = keyValues[i];
            } else {
                adjustedKeyValues[i] = -1;
                adjustedKeyValues[i + 1] = keyValues[i];
            }
        }
        return adjustedKeyValues;
    }

    public OnNumberClickListener getOnItemClickListener() {
        return mOnNumberClickListener;
    }

    public void setOnItemClickListener(OnNumberClickListener onNumberClickListener) {
        this.mOnNumberClickListener = onNumberClickListener;
    }

    public OnDeleteClickListener getOnDeleteClickListener() {
        return mOnDeleteClickListener;
    }

    public void setOnDeleteClickListener(OnDeleteClickListener onDeleteClickListener) {
        this.mOnDeleteClickListener = onDeleteClickListener;
    }

    public CustomizationOptionsBundle getCustomizationOptions() {
        return mCustomizationOptionsBundle;
    }

    public void setCustomizationOptions(CustomizationOptionsBundle customizationOptionsBundle) {
        this.mCustomizationOptionsBundle = customizationOptionsBundle;
    }

    public class NumberViewHolder extends RecyclerView.ViewHolder {
        Button mNumberButton;

        public NumberViewHolder(final View itemView) {
            super(itemView);
            mNumberButton = (Button) itemView.findViewById(R.id.button);
            mNumberButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mOnNumberClickListener != null) {
                        mOnNumberClickListener.onNumberClicked((Integer) v.getTag());
                    }
                }
            });
            mNumberButton.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    if (event.getAction() == MotionEvent.ACTION_DOWN) {
                        if ((Integer) v.getTag() == 0)
                            mNumberButton.setBackgroundResource(R.drawable.btn_0_sel);
                        else
                            mNumberButton.setBackgroundResource(mSelBackgrounds[(Integer) v.getTag() -1]);
                    }
                    if (event.getAction() == MotionEvent.ACTION_UP) {
                       if ((Integer) v.getTag() == 0)
                            mNumberButton.setBackgroundResource(R.drawable.btn_0);
                        else
                            mNumberButton.setBackgroundResource(mBackgrounds[(Integer) v.getTag() -1]);
                    }
                    return false;
                }
            });
        }
    }

    public class DeleteViewHolder extends RecyclerView.ViewHolder {
        //LinearLayout mDeleteButton;
        //ImageView mButtonImage;
        Button mDeleteButton;

        public DeleteViewHolder(final View itemView) {
            super(itemView);
            mDeleteButton = (Button) itemView.findViewById(R.id.button);
            //mButtonImage = (ImageView) itemView.findViewById(R.id.buttonImage);
            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace);

            if (mCustomizationOptionsBundle.isShowDeleteButton() && mPinLength > 0) {
                mDeleteButton.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (mOnDeleteClickListener != null) {
                            mOnDeleteClickListener.onDeleteClicked();
                        }
                    }
                });

                mDeleteButton.setOnLongClickListener(new View.OnLongClickListener() {
                    @Override
                    public boolean onLongClick(View v) {
                        if (mOnDeleteClickListener != null) {
                            mOnDeleteClickListener.onDeleteLongClicked();
                    }
                        return true;
                    }
                });

                mDeleteButton.setOnTouchListener(new View.OnTouchListener() {
                    private Rect rect;

                    @Override
                    public boolean onTouch(View v, MotionEvent event) {
                        if (event.getAction() == MotionEvent.ACTION_DOWN) {
                            /*mButtonImage.setColorFilter(mCustomizationOptionsBundle
                                    .getDeleteButtonPressesColor());*/
                            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace_sel);
                            rect = new Rect(v.getLeft(), v.getTop(), v.getRight(), v.getBottom());
                        }
                        if (event.getAction() == MotionEvent.ACTION_UP) {
                            //mButtonImage.clearColorFilter();
                            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace);
                        }
                        if (event.getAction() == MotionEvent.ACTION_MOVE) {
                            if (!rect.contains(v.getLeft() + (int) event.getX(),
                                    v.getTop() + (int) event.getY())) {
                                //mButtonImage.clearColorFilter();
                                mDeleteButton.setBackgroundResource(R.drawable.btn_backspace_sel);
                            }
                        }
                        return false;
                    }
                });
            }
        }
    }

    public interface OnNumberClickListener {
        void onNumberClicked(int keyValue);
    }

    public interface OnDeleteClickListener {
        void onDeleteClicked();

        void onDeleteLongClicked();
    }
}

File: MyApplication/app/src/main/java/com/martinmarinkovic/myapplication/lockscreen/PinLockAdapter.kt
package com.martinmarinkovic.myapplication.lockscreen

import android.content.Context
import android.graphics.Rect
import android.os.Build
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import androidx.recyclerview.widget.RecyclerView
import com.martinmarinkovic.myapplication.R


class PinLockAdapter(context: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var mContext: Context = context
    private var mKeyValues: IntArray = getAdjustKeyValues(intArrayOf(1, 2, 3, 4, 5, 6, 7, 8, 9, 0))

    companion object {
        private val VIEW_TYPE_NUMBER = 0
        private val VIEW_TYPE_DELETE = 1
        private var mOnNumberClickListener: OnNumberClickListener? = null
        private var mOnDeleteClickListener: OnDeleteClickListener? = null
        private var mCustomizationOptionsBundle: CustomizationOptionsBundle? = null
        private var mPinLength = 0
        private var mBackgrounds: IntArray  = intArrayOf(
            R.drawable.btn_1,
            R.drawable.btn_2,
            R.drawable.btn_3,
            R.drawable.btn_4,
            R.drawable.btn_5,
            R.drawable.btn_6,
            R.drawable.btn_7,
            R.drawable.btn_8,
            R.drawable.btn_9,
            R.drawable.btn_0,
            R.drawable.btn_0
        )
        private var mSelBackgrounds: IntArray  = intArrayOf(
            R.drawable.btn_1_sel,
            R.drawable.btn_2_sel,
            R.drawable.btn_3_sel,
            R.drawable.btn_4_sel,
            R.drawable.btn_5_sel,
            R.drawable.btn_6_sel,
            R.drawable.btn_7_sel,
            R.drawable.btn_8_sel,
            R.drawable.btn_9_sel,
            R.drawable.btn_0_sel,
            R.drawable.btn_0_sel
        )
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val viewHolder: RecyclerView.ViewHolder
        val inflater = LayoutInflater.from(parent.context)
        if (viewType == VIEW_TYPE_NUMBER) {
            val view = inflater.inflate(R.layout.layout_number_item, parent, false)
            viewHolder = NumberViewHolder(view)
        } else {
            val view = inflater.inflate(R.layout.layout_delete_item, parent, false)
            viewHolder = DeleteViewHolder(view)
        }
        return viewHolder
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder.itemViewType == VIEW_TYPE_NUMBER) {
            val vh1: NumberViewHolder = holder as NumberViewHolder
            configureNumberButtonHolder(vh1, position)
        } else if (holder.itemViewType == VIEW_TYPE_DELETE) {
            val vh2: DeleteViewHolder = holder as DeleteViewHolder
            configureDeleteButtonHolder(vh2)
        }
    }

    private fun configureNumberButtonHolder(holder: NumberViewHolder, position: Int) {
        if (holder != null) {
            if (position == 9) {
                holder.mNumberButton.setVisibility(View.GONE)
                holder.mNumberButton.setTag(mKeyValues!![position])
            } else {
                holder.mNumberButton.setBackgroundResource(mBackgrounds!![position])
                //holder.mNumberButton.setText(String.valueOf(mKeyValues[position]));
                holder.mNumberButton.setVisibility(View.VISIBLE)
                holder.mNumberButton.setTag(mKeyValues!![position])
            }
            if (mCustomizationOptionsBundle != null) {
                holder.mNumberButton.setTextColor(mCustomizationOptionsBundle!!.textColor)
                if (mCustomizationOptionsBundle!!.buttonBackgroundDrawable != null) {
                    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
                        holder.mNumberButton.setBackgroundDrawable(
                            mCustomizationOptionsBundle!!.buttonBackgroundDrawable
                        )
                    } else {
                        holder.mNumberButton.setBackground(
                            mCustomizationOptionsBundle!!.buttonBackgroundDrawable
                        )
                    }
                }
                holder.mNumberButton.setTextSize(
                    TypedValue.COMPLEX_UNIT_PX,
                    mCustomizationOptionsBundle!!.textSize.toFloat()
                )
                val params = LinearLayout.LayoutParams(
                    mCustomizationOptionsBundle!!.buttonSize,
                    mCustomizationOptionsBundle!!.buttonSize
                )
                holder.mNumberButton.setLayoutParams(params)
            }
        }
    }

    private fun configureDeleteButtonHolder(holder: DeleteViewHolder?) {
        if (holder != null) {
            if (mCustomizationOptionsBundle!!.isShowDeleteButton && mPinLength > 0) {
                holder.mDeleteButton.setVisibility(View.VISIBLE)

                /*if (mCustomizationOptionsBundle!!.deleteButtonDrawable != null) {
                    holder.mButtonImage.setImageDrawable(mCustomizationOptionsBundle!!.deleteButtonDrawable)
                }
                holder.mButtonImage.setColorFilter(
                    mCustomizationOptionsBundle!!.textColor,
                    PorterDuff.Mode.SRC_ATOP
                )
                val params = LinearLayout.LayoutParams(
                    mCustomizationOptionsBundle!!.deleteButtonSize,
                    mCustomizationOptionsBundle!!.deleteButtonSize
                )
                holder.mButtonImage.setLayoutParams(params)*/

            } else {
                holder.mDeleteButton.setVisibility(View.GONE)
            }
        }
    }

    override fun getItemCount(): Int {
        return 12
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == itemCount - 1) {
            VIEW_TYPE_DELETE
        } else VIEW_TYPE_NUMBER
    }

    fun getPinLength(): Int {
        return mPinLength
    }

    fun setPinLength(pinLength: Int) {
        mPinLength = pinLength
    }

    fun getKeyValues(): IntArray? {
        return mKeyValues
    }

    fun setKeyValues(keyValues: IntArray) {
        mKeyValues = getAdjustKeyValues(keyValues)
        notifyDataSetChanged()
    }

    private fun getAdjustKeyValues(keyValues: IntArray): IntArray {
        val adjustedKeyValues = IntArray(keyValues.size + 1)
        for (i in keyValues.indices) {
            if (i < 9) {
                adjustedKeyValues[i] = keyValues[i]
            } else {
                adjustedKeyValues[i] = -1
                adjustedKeyValues[i + 1] = keyValues[i]
            }
        }
        return adjustedKeyValues
    }

    fun getOnItemClickListener(): OnNumberClickListener? {
        return mOnNumberClickListener
    }

    fun setOnItemClickListener(onNumberClickListener: OnNumberClickListener) {
        mOnNumberClickListener = onNumberClickListener
    }

    fun setOnDeleteClickListener(onDeleteClickListener: OnDeleteClickListener?) {
        mOnDeleteClickListener = onDeleteClickListener
    }

    fun getCustomizationOptions(): CustomizationOptionsBundle? {
        return mCustomizationOptionsBundle
    }

    fun setCustomizationOptions(customizationOptionsBundle: CustomizationOptionsBundle?) {
        mCustomizationOptionsBundle = customizationOptionsBundle
    }

    class NumberViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var mNumberButton: Button = itemView.findViewById<View>(R.id.button) as Button
        init {
            mNumberButton.setOnClickListener { v ->
                mOnNumberClickListener?.onNumberClicked(v.tag as Int)
            }

            mNumberButton.setOnTouchListener { v, event ->
                if (event.action == MotionEvent.ACTION_DOWN) {
                    if (v.tag as Int == 0) mNumberButton.setBackgroundResource(R.drawable.btn_0_sel) else mNumberButton.setBackgroundResource(
                        mSelBackgrounds!![v.tag as Int - 1]
                    )
                }
                if (event.action == MotionEvent.ACTION_UP) {
                    if (v.tag as Int == 0) mNumberButton.setBackgroundResource(R.drawable.btn_0) else mNumberButton.setBackgroundResource(
                        mBackgrounds!![v.tag as Int - 1]
                    )
                }
                false
            }
        }
    }

    class DeleteViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        //LinearLayout mDeleteButton;
        //ImageView mButtonImage;
        var mDeleteButton: Button = itemView.findViewById<View>(R.id.button) as Button

        init {
            //mButtonImage = (ImageView) itemView.findViewById(R.id.buttonImage);
            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace)
            if (mCustomizationOptionsBundle!!.isShowDeleteButton && mPinLength > 0) {
                mDeleteButton.setOnClickListener {
                    mOnDeleteClickListener?.onDeleteClicked()
                }
                mDeleteButton.setOnLongClickListener {
                    mOnDeleteClickListener?.onDeleteLongClicked()
                    true
                }
                mDeleteButton.setOnTouchListener(object : OnTouchListener {
                    private var rect: Rect? = null
                    override fun onTouch(
                        v: View,
                        event: MotionEvent
                    ): Boolean {
                        if (event.action == MotionEvent.ACTION_DOWN) {
                            /*mButtonImage.setColorFilter(mCustomizationOptionsBundle
                                    .getDeleteButtonPressesColor());*/
                            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace_sel)
                            rect = Rect(
                                v.left,
                                v.top,
                                v.right,
                                v.bottom
                            )
                        }
                        if (event.action == MotionEvent.ACTION_UP) {
                            //mButtonImage.clearColorFilter();
                            mDeleteButton.setBackgroundResource(R.drawable.btn_backspace)
                        }
                        if (event.action == MotionEvent.ACTION_MOVE) {
                            if (!rect!!.contains(
                                    v.left + event.x.toInt(),
                                    v.top + event.y.toInt()
                                )
                            ) {
                                //mButtonImage.clearColorFilter();
                                mDeleteButton.setBackgroundResource(R.drawable.btn_backspace_sel)
                            }
                        }
                        return false
                    }
                })
            }
        }
    }

    interface OnNumberClickListener {
        fun onNumberClicked(keyValue: Int)
    }

    interface OnDeleteClickListener {
        fun onDeleteClicked()
        fun onDeleteLongClicked()
    }
}